File: lecture notes/matlab/lec7/run1.m
clear;clc;
T=0.2;

ts=1;
os=10/100;

zeta=-log(os)/sqrt(pi^2+log(os)^2);
wn=4/(ts*zeta);

disp("zeta:"+string(zeta));
disp("wn:"+string(wn));
disp("roots:");
disp(roots([1 2*zeta*wn wn^2]));
zpoles=exp(T*roots([1 2*zeta*wn wn^2]));
disp("z domain:");
disp(zpoles);
coef=conv([1 -zpoles(1)],[1 -zpoles(2)]);
disp(coef);

Gs=tf(1,[1 2]);

Gz=c2d(Gs,T,'zoh');

syms s z;
Gzz=0.1648/(z-0.6703);
syms k;
Tzz=(k*Gzz)/(1+k*Gzz);
Tzz=simplifyFraction(Tzz);
[pzs,pcs]=numden(Tzz);
coef_pcs=coeffs(pcs,z,'all');
coef_pcs=coef_pcs/coef_pcs(1);
prob=coef_pcs==[1 -exp(-(4/ts)*T)];
sol=solve(prob);
sol=double(sol);
kval=sol;

Tz=feedback(kval*Gz,1);
[y,t]=step(Tz);
figure(1);clf;hold on;grid minor;set(gca, 'MinorGridColor', 'k','MinorGridAlpha',1);
xlabel("Zaman(s)");ylabel("y(kT)");title("Basamak Yanıtı");
stem(t,y,'k','LineWidth',2);
print("../../img/"+"lec7_step1.eps",'-depsc','-r150');


figure(2);clf;hold on;grid minor;set(gca, 'MinorGridColor', 'k','MinorGridAlpha',1);
xlabel("Re");ylabel("Im");title("Kök Eğrisi");
zetavec=0:0.1:1;
fx=exp(-zetavec*wn*T).*cos(sqrt(1-zetavec.^2)*wn*T);
fy=exp(-zetavec*wn*T).*sin(sqrt(1-zetavec.^2)*wn*T);
plot(fx,fy,'Color','r','LineWidth',2);
rectangle('Position',[-1,-1,2,2],'Curvature',[1 1]);
r=exp(-(4/ts)*T);
rectangle('Position',[-r,-r,2*r,2*r],'Curvature',[1 1],'EdgeColor','b');
fx=exp(-zeta*wn*T).*cos(sqrt(1-zeta.^2)*wn*T);
fy=exp(-zeta*wn*T).*sin(sqrt(1-zeta.^2)*wn*T);
plot(fx,fy,'kx','LineWidth',2);
plot(0.4493,0,'kx','LineWidth',2);
print("../../img/"+"lec7_rlocus1.eps",'-depsc','-r150');

syms k real;
sol=solve(-0.1648*k+0.6703==-0.4493);
sol=double(sol);
Tz=feedback(sol*Gz,1);
stepinfo(Tz)

[y,t]=step(Tz);
figure(3);clf;hold on;grid minor;set(gca, 'MinorGridColor', 'k','MinorGridAlpha',1);
xlabel("Zaman(s)");ylabel("y(kT)");title("Basamak Yanıtı");
stem(t,y,'k','LineWidth',2);
print("../../img/"+"lec7_step2.eps",'-depsc','-r150');

s=log(-0.4496)/T;
s

File: lecture notes/python/lec7/run1.py
import matplotlib.pyplot as plt
import numpy as np
import sympy as sym
import tbcontrol
import control
############################
# q1
############################
def fun_q1():
    T=0.2
    z,k=sym.symbols("z,k")
    num=0.1648
    den=z-0.6703
    Gz=control.tf(num,np.array(sym.Poly(den,z).all_coeffs(),dtype=np.float64),T)

    os=10/100
    ts=1
    zeta=-np.log(os)/np.sqrt(np.pi**2+np.log(os)**2)
    wn=4/(ts*zeta)
    
    # pole_s=-zeta*wn+1j*wn*np.sqrt(1-zeta**2)
    # pole_z=np.exp(pole_s*T)
    # pdz=(z-pole_z)*(z-np.conj(pole_z))
    # pdz=sym.expand(pdz)
    pole_s=-4/ts
    pole_z=np.exp(pole_s*T)
    pdz=z-pole_z

    Gzz=num/den
    Tzz=(k*Gzz)/(1+k*Gzz)
    Tzz=sym.simplify(Tzz)
    pzz,pcz=sym.fraction(Tzz)
    
    kval=sym.solve(pcz-pdz)
    kval=np.float64(kval[0])
    Fz=control.tf(kval,1,T)
    Tz=control.feedback(control.series(Fz,Gz),1)
    
    t, y = control.step_response(Tz)

    plt.stem(t,y)
    plt.xlim([0,4])
    plt.title("Basamak Yanıtı")
    plt.xlabel("Zaman(s)")
    plt.show()
    
fun_q1()